### PR TITLE
Add Parakeet CTC automatic speech recognition support

### PR DESCRIPTION
**Pull request by justinchuby**
## Summary

Add first-class Mobius support for [`nvidia/parakeet-ctc-1.1b`](https://huggingface.co/nvidia/parakeet-ctc-1.1b), Hugging Face `model_type: parakeet_ctc`.

This is an independent offline CTC ASR architecture and does not duplicate #411, which targets streaming Parakeet RNNT. The new path consumes processor-normalized log-mel features and produces frame-level CTC logits.

## Architecture

- Symmetric 8x depthwise Conv2d FastConformer subsampling
- Transformer-XL relative positional self-attention with learned `bias_u`/`bias_v`
- Macaron feed-forward blocks
- BatchNorm Conformer convolution modules
- 42-layer, hidden-size 1024 encoder and 1025-class Conv1d CTC head
- Feature-input contract: `input_features [B, T, 80]` plus boolean `attention_mask`
- Direct Hugging Face parameter-name alignment; preprocessing only removes BatchNorm `num_batches_tracked`

The reusable component surface also gains grouped/bias-free `Conv1d` and frozen `BatchNorm1d` inference with fp32 accumulation.

## Model and golden integration

- Register `parakeet_ctc` config/model/task extraction
- Add `FeatureCTCAsrTask`
- Add `feature-ctc-asr` YAML/schema and golden generation support
- Pin config and every checkpoint shard to revision `20e63a0fed6aedba145b74b826dbd41df0941730`
- Add exact frame-count enforcement for deterministic CTC L5 goldens
- Add explicit ORT GenAI export rejection for this unsupported feature-input CTC contract

Committed real-audio goldens use nonzero LibriSpeech audio and contain all 114 frame argmax IDs plus the full decoded transcript:

> cauliflower mayonnaise take cold boiled cauliflower break into branches adding salt pepper and vinegar to season

## Validation

GPU: NVIDIA RTX A1000 Laptop GPU, 8 GB; CUDA and `CUDAExecutionProvider` available.

- `python -m pytest tests/build_graph_test.py tests/cli_test.py src/ -q -k "not phi4mm and not apply_weights_unknown" --tb=short -n auto`
  - **3639 passed, 57 skipped**
- `python -m pytest tests/arch_validation_test.py -k parakeet -q --tb=short`
  - **3 passed, 874 deselected**
- `python -m pytest tests/yaml_schema_test.py -q --tb=short`
  - **237 passed**
- Real-checkpoint CUDA parity, fp32:
  - **passed**
  - max absolute difference: **0.000076**
  - mean absolute difference: **0.000014**
- Real-checkpoint CUDA fp16:
  - **passed**
  - exact **114/114** frame IDs
  - exact full transcript
- CUDA L4 checkpoint golden and L5 deterministic CTC transcript/frame golden:
  - **2 passed, 377 deselected**
- `lintrunner -a`
  - **no issues**
- Automated high-confidence code review after fixes:
  - **no significant issues found**

## CLI/runtime/quantization evidence

- fp16 CLI CUDA export loaded and ran with ONNX Runtime CUDA EP
- Runtime produced the exact 114 golden frame IDs and exact full transcript
- Olive 0.13.0 INT4 RTN quantization produced a 751 MB model
- Quantized CUDA inference produced the same full transcript

## Evidence-based limitations and waivers

- **BF16 disabled:** Hugging Face bf16 reference output is correct, and the Mobius graph executes under ORT CUDA after avoiding unsupported kernels, but ORT output collapses to blank CTC predictions. Parakeet therefore defaults safely to fp32 and rejects explicit bf16; fp32 and fp16 are verified. This avoids exposing a configuration that runs without error but returns incorrect ASR results.
- **ORT GenAI / Foundry Local waived:** ORT GenAI 0.15.2 rejects `input_features` in its decoder model schema. This offline feature-input CTC model cannot be represented by the current ORT GenAI decoder contract, so Mobius now fails explicitly rather than producing a misleading package. Foundry Local depends on ORT GenAI and is consequently not applicable until that runtime supports feature-input CTC pipelines.
- **Safetensors package output:** the installed `onnx_ir` failed during safetensors CLI serialization; standard ONNX external-data export succeeded and was used for CUDA runtime and quantization validation.

### CODE REVIEW

**Comment by justinchuby** [comment]
### CI baseline / unrelated-failure analysis (2026-08-10)

The remaining failures in CI run [31418606770](https://github.com/onnxruntime/mobius/actions/runs/31418606770) are outside the Parakeet change surface:

- Dedicated L3 job [93553773549](https://github.com/onnxruntime/mobius/actions/runs/31418606770/job/93553773549) fails only `bamba_0` and `bamba_1`. The reported values are `max_abs_diff=0.001375`, `mean_abs_diff=0.000201`, cosine `0.999997`, argmax match true; and `max_abs_diff=0.001662`, `mean_abs_diff=0.000172`, cosine `0.999996`, argmax match true.
- Every platform job (Ubuntu and Windows, Python 3.11/3.12/3.13) has the same two Bamba failures and no Parakeet-specific failure. The six job IDs are `93553709286`, `93553709345`, `93553709364`, `93553709475`, `93553709369`, and `93553709299`.
- Integration(fast) job [93553709271](https://github.com/onnxruntime/mobius/actions/runs/31418606770/job/93553709271) fails only `test_deepseek_v2_lite_prefill_logits_match` and `test_deepseek_non_mla_decoder_prefill_logits_match` (`2 failed, 16 passed, 4 skipped, 100 deselected`).
- This PR does not modify Bamba or DeepSeek model/test code. No unrelated tolerance or model changes are included to mask those failures.

The exact base SHA is `78437cd01b2391a53c757bd3d06850ff493494ae`. Its main CI run [31230202537](https://github.com/onnxruntime/mobius/actions/runs/31230202537) is itself red: L4 and L5 each time out on the unrelated `olmoe-1b-7b` and `qwen1_5-moe` cases. This confirms that the base was not fully green before the Parakeet commit; it does not claim that the older base run reproduced the later Bamba/DeepSeek failures.

Parakeet-specific L1 is green in job [93553773623](https://github.com/onnxruntime/mobius/actions/runs/31418606770/job/93553773623). Local CUDA L4/L5 and real-checkpoint fp32/fp16 validations are documented in the PR body; GitHub L4/L5 are still running in this CI run and will be checked to completion.

**Comment by justinchuby** [comment]
### Hosted Parakeet golden follow-up

The Parakeet-specific validation is now complete on the hosted GPU runner at head `3262860d8af4bb63aa4c9fb098f7dacc42d86d0a`:

- L1 smoke job `93553773623`: **passed**.
- Targeted L4 run [31485270551](https://github.com/onnxruntime/mobius/actions/runs/31485270551), job [93759094978](https://github.com/onnxruntime/mobius/actions/runs/31485270551/job/93759094978): **passed**; the checkpoint test step completed in 2m04s.
- Targeted L5 run [31485270592](https://github.com/onnxruntime/mobius/actions/runs/31485270592), job [93759095610](https://github.com/onnxruntime/mobius/actions/runs/31485270592/job/93759095610): **passed**; the generation test step completed in 2m16s.

The earlier aggregate L4/L5 failures were infrastructure-only. The first two attempts exhausted the self-hosted runner disk because `HF_HOME` was changed after `huggingface_hub` had captured process-global cache constants, and Xet chunks were not redirected or cleaned. Commit `0daf582` now patches the hub/assets/Xet runtime cache constants per test and eagerly deletes each cache at teardown; local validation confirmed both Parakeet CUDA goldens pass and both test caches are absent afterward. The subsequent aggregate jobs no longer exhausted disk, but hit the existing one-hour all-model workflow ceiling, consistent with the already-red base golden runs.

Commit `3262860` exposes the existing affected-model input to manual dispatch, allowing the successful hosted Parakeet-only evidence above without changing, skipping, or loosening unrelated Bamba/DeepSeek tests.

**Review by justinchuby — COMMENTED** [review]
Correctness review: two findings.

`src/mobius/_builder.py:539`: the new `revision` argument is dropped on the diffusers fallback. If `AutoConfig` fails and this path is selected, the pipeline index/configs/weights come from the repository default revision rather than the requested commit. Please thread `revision` through `build_diffusers_pipeline` and its Hub downloads.

File: src/mobius/components/_parakeet_audio.py [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+            bias=config.attention_bias,
+        )
+        self.linear2 = Linear(
+            config.intermediate_size,
+            config.hidden_size,
+            bias=config.attention_bias,
+        )
+
+    def forward(self, op: OpBuilder, hidden_states: ir.Value) -> ir.Value:
+        return self.linear2(op, op.Swish(self.linear1(op, hidden_states)))
```
Comment (justinchuby) [inline src/mobius/components/_parakeet_audio.py]: `hidden_act` is part of the Parakeet config contract, but both feed-forward blocks and the convolution module hardcode Swish (the latter at line 377). A valid checkpoint using another configured activation will build successfully but produce incorrect logits. Please resolve the configured activation once and use it in both paths; a non-SiLU parity case would prevent regression.

File: tests/e2e_golden_test.py [hunk 2]
```diff
@@ -408,6 +408,7 @@ def _build_model_package(case: GoldenTestCase) -> ModelPackage:
         task = reg.task or getattr(module_class, "default_task", None)
     return build(
         case.model_id,
+        revision=case.revision,
         task=task,
         module_class=module_class,
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply tests/e2e_golden_test.py:452]: Fixed in 612f3eb. Every Hub-backed `from_pretrained` call in the golden harness now receives `case.revision`, including processors, feature extractors, configs, generation configs, and Whisper fallbacks. Added a static regression audit so future unpinned artifact loads fail tests.

File: src/mobius/components/_parakeet_audio.py (hunk 1 above)
Comment (justinchuby) [reply src/mobius/components/_parakeet_audio.py]: Fixed in 612f3eb. Both Parakeet Macaron FFNs and the convolution module now use `get_activation(config.hidden_act)`. Added a non-SiLU graph test verifying all three configurable sites become ReLU and no Swish remains.

**Comment by justinchuby** [comment]
Also fixed the review-summary finding in 612f3eb: `revision` is now propagated through the diffusers fallback, pipeline index, component configs, weight indexes/files, and parallel shard downloads, with targeted no-network tests.

**Comment by justinchuby** [comment]
Updated PR #471 to current `main` (`1e6f61f`) in merge commit `dd1122e` and revalidated Parakeet's optimized graph.

Optimization/fusion evidence:
- Tiny 2-layer graph: **406 raw nodes -> 270 after Mobius optimization -> 219 after ORT CUDA optimization**.
- Mobius graph retains **2 opset-24 `Attention`**, **8 `SkipLayerNormalization`**, **6 native opset-24 `Swish`**, and **2 `BatchNormalization`** nodes.
- Frozen `BatchNorm1d` now emits standard `BatchNormalization`; ORT CUDA folds both BatchNorm nodes into their adjacent convolutions (**2 -> 0** in the optimized graph).
- ORT CUDA lowers the six native Swish nodes to six `com.microsoft::QuickGelu` kernels with alpha 1.
- Relative-position attention requires an explicit additive bias, so Flash Attention is not eligible; fused ONNX `Attention` remains the correct representation and FP16 can use MEA when sequence alignment permits.

Post-merge validation:
- Full non-integration suite: **3867 passed, 58 skipped**.
- Merged targeted tests: **50 passed**; fused graph tests: **15 passed**; Parakeet schema/architecture selection: **4 passed**.
- `lintrunner -a`: **no issues**.
- Real FP32 CUDA parity: max absolute difference **0.000137**, mean **0.000014**.
- Real FP16 CUDA: exact **114 frame IDs** and exact transcript.
- CUDA L4/L5 Parakeet goldens: **2 passed**.
- Focused automated code review found no significant issues.

**Comment by justinchuby** [comment]
Final main refresh completed in `9abd39c`: PR #471 now includes `main@ed0319d` (Moonshine ASR) while retaining Parakeet CTC and the graph-fusion changes.

The only new integration conflict was the shared Whisper/Moonshine `Conv1d`: the resolution preserves the existing positional `bias` API and adds grouped weights/`group` dispatch for Parakeet depthwise convolution. Automated review caught and the branch now covers positional-bias compatibility explicitly.

Final post-merge evidence:
- Full non-integration suite from this worktree: **3880 passed, 58 skipped**.
- Focused Parakeet/Moonshine/Whisper/BatchNorm suite: **49 passed**.
- `lintrunner -a`: **no issues**.
- Final tiny Parakeet optimized graph: **188 nodes**, including exactly **1 `Attention`**, **1 `BatchNormalization`**, **3 `Swish`**, **4 `SkipLayerNormalization`**, and **0 `Sqrt`** nodes.
- Hosted replacement **Lint: passed**.
- Hosted replacement **Architecture Diff: passed**.

Earlier real-checkpoint CUDA FP32/FP16 and hosted Parakeet-only L4/L5 evidence remains applicable because the latest merge did not alter Parakeet math; its only shared-component resolution preserves the already-validated grouped Conv1d graph and is covered by the graph tests above.
